store/postgres/migration.go

```diff
@@ -0,0 +1,232 @@
+		return fmt.Errorf("optimus version [%s] should be higher or equal than existing [%s]", m.incomingOptimusVersion, latestStep.CurrentOptimusVersion)
+	}
+	if m.incomingOptimusVersion == latestStep.CurrentOptimusVersion {
+		m.logger.Warn(fmt.Sprintf("migration up is skipped because optimus version [%s] is the same as current one", m.incomingOptimusVersion))
+		return nil
+	}
+
+	migrationClient, migrationClientCleanup, err := m.newMigrationClient()
+	if err != nil {
+		return fmt.Errorf("error initializing migration client: %w", err)
+	}
+	defer migrationClientCleanup()
+
+	if err := migrationClient.Up(); err != nil && !errors.Is(err, migrate.ErrNoChange) {
+		return fmt.Errorf("error executing migration up: %w", err)
+	}
+	newVersion, _, err := migrationClient.Version()
+	if err != nil {
+		return fmt.Errorf("error getting current migration version: %w", err)
+	}
+
+	newMigrationVersion := &migrationStep{
+		CurrentOptimusVersion:   m.incomingOptimusVersion,
+		CurrentMigrationVersion: newVersion,
+		PreviousOptimusVersion:  latestStep.CurrentOptimusVersion,
+		CreatedAt:               time.Now(),
+	}
+	return m.addMigrationStep(ctx, dbClient, newMigrationVersion)
+}
+
+func (m *migration) Rollback(ctx context.Context) error {
+	dbClient, dbClientCleanup, err := m.newDBClient()
+	if err != nil {
+		return fmt.Errorf("error initializing db client: %w", err)
+	}
+	defer dbClientCleanup()
+
+	if err := dbClient.WithContext(ctx).AutoMigrate(&migrationStep{}); err != nil {
+		return fmt.Errorf("error setting up migration_steps: %w", err)
+	}
+
+	latestStep, err := m.getLatestMigrationStep(ctx, dbClient)
+	if err != nil {
+		return err
+	}
+	if m.incomingOptimusVersion != latestStep.CurrentOptimusVersion {
+		return fmt.Errorf("expecting optimus with version [%s] but got [%s]", latestStep.CurrentOptimusVersion, m.incomingOptimusVersion)
+	}
+
+	previousMigrationVersion, err := m.getMigrationVersion(ctx, dbClient, latestStep.PreviousOptimusVersion)
+	if err != nil {
+		return err
+	}
+	if previousMigrationVersion == 0 {
+		m.logger.Warn("migration rollback is skipped because previous migration version is not registered")
+		return nil
+	}
+
+	migrationClient, migrationClientCleanup, err := m.newMigrationClient()
+	if err != nil {
+		return fmt.Errorf("error initializing migration client: %w", err)
+	}
+	defer migrationClientCleanup()
+
+	if err := migrationClient.Migrate(previousMigrationVersion); err != nil && !errors.Is(err, migrate.ErrNoChange) {
+		return fmt.Errorf("error migrating to version [%d]: %w", previousMigrationVersion, err)
+	}
+	return m.removeMigrationStep(ctx, dbClient, latestStep)
+}
+
+func (m *migration) newMigrationClient() (migrationClient *migrate.Migrate, cleanup func(), err error) {
+	path := "migrations"
+	sourceDriver, sourceDriverErr := iofs.New(migrationFs, path)
+	if sourceDriverErr != nil {
+		err = fmt.Errorf("error initializing source driver: %w", sourceDriverErr)
+		return
```

Comment:
Avoid naked return, multiple return values are hard to keep in head.
Also assigning to `err` in case of actual error.

Reply:
since we have more than two returns, using named returns is so that the caller could understand in a glance what each return is about. in some cases, they are good to implement. however, I do agree that in the current case, where the function is a bit long, it's not a good approach. I already updated it.